src/Nethermind/Nethermind.Merge.Plugin/Synchronization/MergeBlockDownloader.cs

```diff
@@ -104,33 +104,54 @@ public class MergeBlockDownloader : BlockDownloader
 
             bool HasMoreToSync()
                 => currentNumber < _blockTree.BestKnownBeaconNumber &&
-                   bestPeer.HeadNumber > _blockTree.BestKnownNumber;
+                   bestPeer.HeadNumber > currentNumber;
+
+            bool foundBeaconBlocks = false;
 
             while (HasMoreToSync())
             {
                 if (_logger.IsDebug)
                     _logger.Debug($"Continue full sync with {bestPeer} (our best {_blockTree.BestKnownNumber})");
+
+                long upperDownloadBoundary = _blockTree.BestKnownBeaconNumber;
+                long blocksLeft = upperDownloadBoundary - currentNumber;
+                int headersToProcess = (int)Math.Min(blocksLeft + 1, _syncBatchSize.Current);
+                if (headersToProcess <= 1)
+                {
+                    break;
+                }
                 
-                int headersToRequest = Math.Min(_syncBatchSize.Current, bestPeer.MaxHeadersPerRequest());
+                int requestSize = Math.Min(headersToProcess, bestPeer.MaxHeadersPerRequest());
+
                 if (_logger.IsTrace)
                     _logger.Trace(
-                        $"Full sync request {currentNumber}+{headersToRequest} to peer {bestPeer} with {bestPeer.HeadNumber} blocks. Got {currentNumber} and asking for {headersToRequest} more.");
+                        $"Full sync request {currentNumber}+{requestSize} to peer {bestPeer} with {bestPeer.HeadNumber} blocks. Got {currentNumber} and asking for {requestSize} more.");
 
                 if (cancellation.IsCancellationRequested) return blocksSynced; // check before every heavy operation
-                Block[]? blocks = null;
-                TxReceipt[]?[]? receipts = null;
+                Block[]? blocks;
+                TxReceipt[]?[]? receipts;
                 if (_logger.IsTrace)
                     _logger.Trace(
-                        $"Downloading blocks from peer. CurrentNumber: {currentNumber}, BeaconPivot: {_beaconPivot.PivotNumber}, BestPeer: {bestPeer}, HeaderToRequest: {headersToRequest}");
+                        $"Downloading blocks from peer. CurrentNumber: {currentNumber}, BeaconPivot: {_beaconPivot.PivotNumber}, BestPeer: {bestPeer}, RequestSize: {requestSize}");
                 
-                BlockHeader[]? headers = _chainLevelHelper.GetNextHeaders(headersToRequest);
+                BlockHeader[] headers = _chainLevelHelper.GetNextHeaders(requestSize);
                 if (headers == null || headers.Length == 0)
                     break;
+                
                 BlockDownloadContext context = new(_specProvider, bestPeer, headers, downloadReceipts,
-                    _receiptsRecovery);
-
+                _receiptsRecovery);
```

Comment:
NP: Indentation.

Reply:
BTW, is there a linter setup for this?

Reply:
we will have a very big PR when the whole codebase goes through linter 😅